Add log lines to the test report for the failed test cases

mgalka opened the pull request:
Modify test report template to display the log lines stored along with the test cases when the test failed. The number of the log lines in a single fragment is limited to 10. If there are more log lines in
a fragment they are truncated and an appropriate message is added.

gctucker commented:
Here's a sample:
```
  bootrr - 84 tests: 81  PASS, 3 FAIL, 0 SKIP
    * cros-ec-sensors-accel0-probed:
        Test case HTML log: https://storage.staging.kernelci.org//kernelci/staging.kernelci.org/staging-20200429.2/arm64/defconfig/gcc-8/lab-collabora/baseline-rk3399-gru-kevin.html#L1406
        never passed
        Log lines:
            2020-04-29 14:32:33.434000 <8>[   20.726199] <LAVA_SIGNAL_TESTCASE TEST_CASE_ID=cros-ec-sensors-driver-present RESULT=pass>
            2020-04-29 14:32:34.445000 /lava-2375898/1/../bin/lava-test-case 
    * cros-ec-sensors-accel1-probed:
        Test case HTML log: https://storage.staging.kernelci.org//kernelci/staging.kernelci.org/staging-20200429.2/arm64/defconfig/gcc-8/lab-collabora/baseline-rk3399-gru-kevin.html#L1410
        never passed
        Log lines:
            2020-04-29 14:32:34.456000 <8>[   21.748061] <LAVA_SIGNAL_TESTCASE TEST_CASE_ID=cros-ec-sensors-accel0-probed RESULT=fail>
            2020-04-29 14:32:35.468000 /lava-2375898/1/../bin/lava-test-case 
    * cros-ec-sensors-gyro0-probed:
        Test case HTML log: https://storage.staging.kernelci.org//kernelci/staging.kernelci.org/staging-20200429.2/arm64/defconfig/gcc-8/lab-collabora/baseline-rk3399-gru-kevin.html#L1415
        never passed
        Log lines:
            2020-04-29 14:32:35.479000 <8>[   22.770622] <LAVA_SIGNAL_TESTCASE TEST_CASE_ID=cros-ec-sensors-accel1-probed RESULT=fail>
            2020-04-29 14:32:36.490000 /lava-2375898/1/../bin/lava-test-case
            2020-04-29 14:32:36.501000 <8>[   23.793906] <LAVA_SIGNAL_TESTCASE TEST_CASE_ID=cros-ec-sensors-gyro0-probed RESULT=fail> 
```
We should be able to filter out all these lines as they are artificially added by LAVA.  Other than that, the email template seems to be doing its job fine.

Another sample:
```
  1 tests: 0 PASS, 1 FAIL, 0 SKIP
    * login:
        new failure (last pass: staging-20200428.3)
        Log lines:    
```
There were no log lines in this case, so I think "Log lines:" can be dropped when that happens.  In fact, it may be better to just have a blank line between the test result details and the log lines, like this:

```
    * cros-ec-sensors-gyro0-probed:
        Test case HTML log: https://storage.staging.kernelci.org//kernelci/staging.kernelci.org/staging-20200429.2/arm64/defconfig/gcc-8/lab-collabora/baseline-rk3399-gru-kevin.html#L1415
        never passed

            2020-04-29 14:32:35.479000  <8>[   22.770622] <LAVA_SIGNAL_TESTCASE TEST_CASE_ID=cros-ec-sensors-accel1-probed RESULT=fail>
            2020-04-29 14:32:36.490000  /lava-2375898/1/../bin/lava-test-case
            2020-04-29 14:32:36.501000  <8>[   23.793906] <LAVA_SIGNAL_TESTCASE TEST_CASE_ID=cros-ec-sensors-gyro0-probed RESULT=fail>
```
Also having 2 spaces between the timestamps and the log line message would make it easier to read (as in the example above).  That's what is done in the HTML logs too.

mgalka commented:
> There were no log lines in this case, so I think "Log lines:" can be dropped when that happens. In fact, it may be better to just have a blank line between the test result details and the log lines, like this:

> Also having 2 spaces between the timestamps and the log line message would make it easier to read (as in the example above). That's what is done in the HTML logs too.

I replaced the "Log lines:" string with a blank line and added 2 spaces between the timestamps and the log line message.

khilman commented:
The example reports from @gctucker show both a link to line number the HTML log (because Khouloud's PR https://github.com/kernelci/kernelci-backend/pull/207 is also in staging), which leads to the question do we want/need both a link to the log and the log lines themselves?

I think the answer is yes, but not always.  e.g. for noisy failing tests, this is going to lead to very noisy email reports.  I wonder if we should have some logic here that only shows the log lines if there are fewer than N (e.g. 5), and otherwise just lists the number?

gctucker commented:
Yes I agree, we should have some logic in the template to either only show a maximum number of lines and some "..." when it goes over that, or no lines at all.  The link should always be useful to directly find the test case output in the full log.

gctucker commented:
I think there's one blank line too many:
```
  dmesg - 3 tests: 2  PASS, 1 FAIL, 0 SKIP
    * alert:
        Test case HTML log: https://storage.staging.kernelci.org//kernelci/staging.kernelci.org/staging-20200430.2/arm/multi_v7_defconfig/clang-9/lab-collabora/baseline-exynos4412-odroidx2.html#L858
        never passed
        1 lines

        
            2020-04-30 23:37:32.007000  kern  :alert : L2C: platform provided aux values permit register corruption.   
```
and there is also an extra blank line when there are no log lines.  So it's probably missing a `-` sign in a tag template.

mgalka commented:
Changes:

- Limit log fragment size to 10 lines
- Add appropriate message when the fragment is truncated
- Remove extra blank lines

Rendered template example:

```
 1 tests: 0 PASS, 1 FAIL, 0 SKIP
    * login:
        never passed      

  dmesg - 3 tests: 2  PASS, 1 FAIL, 0 SKIP
    * crit:
        never passed
        0 lines

            2020-05-12 14:50:59.513000  / #
            2020-05-12 14:50:59.521000   
            2020-05-12 14:50:59.625000  / # #
            2020-05-12 14:50:59.629000  <4>[    3.038227] urandom_read: 2 callbacks suppressed
            2020-05-12 14:50:59.631000  <5>[    3.038227] random: udevd: uninitialized urandom read (16 bytes read)
            2020-05-12 14:50:59.633000  <6>[    3.051614] input: ImExPS/2 Generic Explorer Mouse as /devices/platform/i8042/serio1/input/input3
            2020-05-12 14:50:59.639000  <27>[    3.062912] udevd[1076]: specified group 'kvm' unknown
            2020-05-12 14:50:59.644000  #
            2020-05-12 14:50:59.753000  / #mkdir /lava-488
            2020-05-12 14:50:59.855000  mount /dev/disk/by-uuid/b73ebbf5-46aa-4022-b08e-e699c5704b22 -t ext2 /lava-488
            ... (31 line(s) more)
```

gctucker commented:
Looks good, and thanks for sharing an example while staging is frozen this week.

Maybe the plural case can be handled with the Jinja2 pluralize filter in `line(s)`?
https://jinja2-pluralize.readthedocs.io/en/latest/readme.html

mgalka commented:
> Looks good, and thanks for sharing an example while staging is frozen this week.
> 
> Maybe the plural case can be handled with the Jinja2 pluralize filter in `line(s)`?
> https://jinja2-pluralize.readthedocs.io/en/latest/readme.html

"pluralize" is not a built-in filter. Do you think it's worth installing another package to use it just for this?

gctucker commented:
OK maybe not worth it then.  We can review email templates once we've gone through all the functional changes and refactoring and then revisit the situation.

gctucker commented:
Sample report:
```
platform                     | arch  | lab          | compiler | defconfig | results
-----------------------------+-------+--------------+----------+-----------+--------
bcm2837-rpi-3-b              | arm64 | lab-baylibre | clang-9  | defconfig | 4/5    

  Details:     https://staging.kernelci.org/test/plan/id/5ecc05dc13b7f7a6c077baf6

  Results:     4 PASS, 1 FAIL, 0 SKIP
  Full config: defconfig
  Compiler:    clang-9 (clang version 9.0.1-+20191211082622+c1a0a213378-1~exp1~20191211193200.102 )
  Plain log:   https://storage.staging.kernelci.org//kernelci/staging.kernelci.org/staging-20200525.2/arm64/defconfig/clang-9/lab-baylibre/baseline-bcm2837-rpi-3-b.txt
  HTML log:    https://storage.staging.kernelci.org//kernelci/staging.kernelci.org/staging-20200525.2/arm64/defconfig/clang-9/lab-baylibre/baseline-bcm2837-rpi-3-b.html
  Rootfs:      http://storage.staging.kernelci.org/images/rootfs/buildroot/kci-2019.02-12-ga2fb3c8af5e6/arm64/baseline/rootfs.cpio.gz 

  * baseline.dmesg.crit: https://staging.kernelci.org/test/case/id/5ecc05dc13b7f7a6c077baf9
      failing since 3 days (last pass: staging-20200519.2, first fail: staging-20200522.2)
      2 lines

    2020-05-25 17:52:11.938000  / # 
    2020-05-25 17:52:11.949000  
    2020-05-25 17:52:12.053000  / # #
    2020-05-25 17:52:12.059000  #
    2020-05-25 17:52:13.318000  / # export SHELL=/bin/sh
    2020-05-25 17:52:13.329000  export SHELL=/bin/sh[   27.581865] hwmon hwmon1: Undervoltage detected!
    2020-05-25 17:52:13.329000  
    2020-05-25 17:52:14.889000  / # . /lava-36202/environment
    2020-05-25 17:52:14.903000  . /lava-36202/environment
    2020-05-25 17:52:17.731000  / # /lava-36202/bin/lava-test-runner /lava-36202/0
    ... (10 line(s) more)
```
Apart from the LAVA noise in the log, I think this is fine from an email template point of view.
